// wouter/cjs/react-deps.js
'use strict';

var react = require('react');
var useSyncExternalStore = require('./use-sync-external-store');

// Copied from:
// https://github.com/facebook/react/blob/main/packages/shared/ExecutionEnvironment.js
const canUseDOM = !!(
  typeof window !== "undefined" &&
  typeof window.document !== "undefined" &&
  typeof window.document.createElement !== "undefined"
);

// Copied from:
// https://github.com/reduxjs/react-redux/blob/master/src/utils/useIsomorphicLayoutEffect.ts
// "React currently throws a warning when using useLayoutEffect on the server.
// To get around it, we can conditionally useEffect on the server (no-op) and
// useLayoutEffect in the browser."
const useIsomorphicLayoutEffect = canUseDOM
  ? react.useLayoutEffect
  : react.useEffect;

// useInsertionEffect is already a noop on the server.
// See: https://github.com/facebook/react/blob/main/packages/react-server/src/ReactFizzHooks.js
const useInsertionEffect =
  react.useInsertionEffect || useIsomorphicLayoutEffect;

// Userland polyfill while we wait for the forthcoming
// https://github.com/reactjs/rfcs/blob/useevent/text/0000-useevent.md
// Note: "A high-fidelty polyfill for useEvent is not possible because
// there is no lifecycle or Hook in React that we can use to switch
// .current at the right timing."
// So we will have to make do with this "close enough" approach for now.
const useEvent = (fn) => {
  const ref = react.useRef([fn, (...args) => ref[0](...args)]).current;
  // Per Dan Abramov: useInsertionEffect executes marginally closer to the
  // correct timing for ref synchronization than useLayoutEffect on React 18.
  // See: https://github.com/facebook/react/pull/25881#issuecomment-1356244360
  useInsertionEffect(() => {
    ref[0] = fn;
  });
  return ref[1];
};

Object.defineProperty(exports, 'Fragment', {
  enumerable: true,
  get: function () { return react.Fragment; }
});
Object.defineProperty(exports, 'cloneElement', {
  enumerable: true,
  get: function () { return react.cloneElement; }
});
Object.defineProperty(exports, 'createContext', {
  enumerable: true,
  get: function () { return react.createContext; }
});
Object.defineProperty(exports, 'createElement', {
  enumerable: true,
  get: function () { return react.createElement; }
});
Object.defineProperty(exports, 'forwardRef', {
  enumerable: true,
  get: function () { return react.forwardRef; }
});
Object.defineProperty(exports, 'isValidElement', {
  enumerable: true,
  get: function () { return react.isValidElement; }
});
Object.defineProperty(exports, 'useContext', {
  enumerable: true,
  get: function () { return react.useContext; }
});
Object.defineProperty(exports, 'useState', {
  enumerable: true,
  get: function () { return react.useState; }
});
Object.defineProperty(exports, 'useSyncExternalStore', {
  enumerable: true,
  get: function () { return useSyncExternalStore.useSyncExternalStore; }
});
exports.useEvent = useEvent;
exports.useInsertionEffect = useInsertionEffect;
exports.useIsomorphicLayoutEffect = useIsomorphicLayoutEffect;
